[SPARK-6766][Streaming] Fix issue about StreamingListenerBatchSubmitted and StreamingListenerBatchStarted

zsxwing opened the pull request:
This PR includes:
1. Send `StreamingListenerBatchSubmitted` when `JobSet` is submitted
2. Fix `StreamingListenerBatchStarted.batchInfo.processingStartTime`
3. Fix a type: `completedaBatchInfos` -> `completedBatchInfos`

zsxwing committed:
Send StreamingListenerBatchSubmitted when JobSet is submitted; fix StreamingListenerBatchStarted.batchInfo.processingStartTime; fix a typo

zsxwing commented:
cc @tdas

SparkQA commented:
[Test build #29849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29849/consoleFull) for   PR 5414 at commit [`493f978`](https://github.com/apache/spark/commit/493f9783fd34329f3ca7f596687d0daf7b6dbf27).

jerryshao commented:
Good catch, LGTM :).

SparkQA commented:
[Test build #29849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29849/consoleFull) for   PR 5414 at commit [`493f978`](https://github.com/apache/spark/commit/493f9783fd34329f3ca7f596687d0daf7b6dbf27).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29849/
Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala:
```diff
@@ -33,7 +33,7 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
   private val waitingBatchInfos = new HashMap[Time, BatchInfo]
   private val runningBatchInfos = new HashMap[Time, BatchInfo]
-  private val completedaBatchInfos = new Queue[BatchInfo]
+  private val completedBatchInfos = new Queue[BatchInfo]
```
Good catch. OMG!

tdas commented:
This looks good to me at first glance, but I am concerned that such a thing went unnoticed for so long despite unit tests in StreamingListenerSuite. Let me think whether we can improve the tests. If so, we definitely should add tests as part of this PR so that such a thing cannot happen again.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala:
```diff
@@ -135,9 +136,13 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
   private def handleJobStart(job: Job) {
     val jobSet = jobSets.get(job.time)
     if (!jobSet.hasStarted) {
+      jobSet.handleJobStart(job)
```
I think this can be refactored to dedup lines 139 and 144.

jerryshao replied:
I think it is hard to dedup cause of ordering call.

zsxwing replied:
Yes, `handleJobStart` will set `jobSet.hasStarted` to true.

tdas replied:
How about 
```
val isFirstJobOfJobSet = !jobSet.hasStarted
jobSet.handleJobStart(job)
if (isFirstJobOfJobSet) {
    listenerBus.post(StreamingListenerBatchStarted(jobSet.toBatchInfo))
}
```

zsxwing replied:
Good idea. Updated.

tdas commented:
Overall, I realized that there is not tests for StreamingJobProgressListener, which is my bad. :( We really need to add some tests for the StreamingJobProgressListener which tests some basic aspects like this. Could you add a `StreamingJobProgressListenerSuite` ?

zsxwing commented:
> Could you add a StreamingJobProgressListenerSuite ?

OK. I will add some basic tests for StreamingJobProgressListener

zsxwing committed:
Refactor as per TD's suggestion

zsxwing committed:
Add unit tests for SPARK-6766

zsxwing committed:
Add StreamingJobProgressListenerSuite to test StreamingJobProgressListener

zsxwing commented:
Added some tests to StreamingListenerSuite for this issue, and added StreamingJobProgressListenerSuite for some basic tests of StreamingJobProgressListener.

SparkQA commented:
[Test build #29935 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29935/consoleFull) for   PR 5414 at commit [`79b4fed`](https://github.com/apache/spark/commit/79b4fedad774c5617425349540c9ae83c8996df9).

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingListenerSuite.scala:
```diff
@@ -88,6 +88,58 @@ class StreamingListenerSuite extends TestSuiteBase with Matchers {
     }
   }
 
+  test("SPARK-6766: batch info should be submitted") {
```
How long does these two units tests take? I wonder whether it is best (in the interest of minimizing test run times) to collapse all these two unit tests into the earlier "batch info reporting".

zsxwing replied:
[info] StreamingListenerSuite:
[info] - batch info reporting (4 seconds, 976 milliseconds)
[info] - receiver info reporting (847 milliseconds)
[info] - SPARK-6766: batch info should be submitted (667 milliseconds)
[info] - SPARK-6766: processingStartTime of batch info should not be None when starting (691 milliseconds)
[info] ScalaTest
[info] Run completed in 8 seconds, 996 milliseconds.
[info] Total number of tests run: 4
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.

zsxwing replied:
This was my last test result.

tdas replied:
Hmm, each of the two new unit tests is taking .7 seconds. Not good. Best to merge them into the first unit test. 
The reason its taking time is that its starting a new SparkContext every time, and running batches after that.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala:
```diff
@@ -0,0 +1,140 @@
+
+  override def batchDuration = Milliseconds(100)
+
+  test("onBatchSubmitted") {
```
Same comment as above. How long do these unit tests take? If they are in total taking less than a second or two, its fine. Otherwise it might be best to collapse a few of them into a smaller number of tests.

zsxwing replied:
[info] StreamingJobProgressListenerSuite:
[info] - onBatchSubmitted (3 seconds, 172 milliseconds)
[info] - onBatchStarted (243 milliseconds)
[info] - onBatchCompleted (203 milliseconds)
[info] - retain completed batch (162 milliseconds)
[info] - onReceiverStarted (142 milliseconds)
[info] - onReceiverError (306 milliseconds)
[info] - onReceiverStopped (164 milliseconds)
[info] ScalaTest
[info] Run completed in 6 seconds, 319 milliseconds.
[info] Total number of tests run: 7
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 7, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.

zsxwing replied:
This was my last test result.

tdas replied:
Same concern. Best to merge these tests together into 1 or 2 unit tests. Furthermore this suite is going to have a lot more unit tests any ways when we join job and batch information. So lets merge these unit tests together into 1 or 2 tests. Will save lot of testing time.

SparkQA commented:
**[Test build #29935 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29935/consoleFull)**     for PR 5414 at commit [`79b4fed`](https://github.com/apache/spark/commit/79b4fedad774c5617425349540c9ae83c8996df9)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29935/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
[Test build #29937 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29937/consoleFull) for   PR 5414 at commit [`79b4fed`](https://github.com/apache/spark/commit/79b4fedad774c5617425349540c9ae83c8996df9).

SparkQA commented:
[Test build #29937 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29937/consoleFull) for   PR 5414 at commit [`79b4fed`](https://github.com/apache/spark/commit/79b4fedad774c5617425349540c9ae83c8996df9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29937/
Test PASSed.

zsxwing committed:
Combine unit tests

SparkQA commented:
[Test build #29985 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29985/consoleFull) for   PR 5414 at commit [`ca0955b`](https://github.com/apache/spark/commit/ca0955b492279414332d38a2b39107ad5e05598a).

zsxwing commented:
Collapsed the new unit tests together.

SparkQA commented:
[Test build #29985 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29985/consoleFull) for   PR 5414 at commit [`ca0955b`](https://github.com/apache/spark/commit/ca0955b492279414332d38a2b39107ad5e05598a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29985/
Test PASSed.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingListenerSuite.scala:
```diff
@@ -46,6 +46,8 @@ class StreamingListenerSuite extends TestSuiteBase with Matchers {
     val collector = new BatchInfoCollector
     ssc.addStreamingListener(collector)
     runStreams(ssc, input.size, input.size)
+    ssc.awaitTerminationOrTimeout(5000) should be (true)
```
Why this line? Run streams should have already stopped the ssc, isnt it?

zsxwing replied:
Sorry that I misread `runStreams.` This line is not necessary.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala:
```diff
@@ -0,0 +1,106 @@
+    listener.onBatchStarted(StreamingListenerBatchStarted(batchInfoStarted))
+    listener.runningBatches should be(List(batchInfoStarted))
+    listener.waitingBatches should be(Nil)
+    listener.numTotalReceivedRecords should be(600)
```
Shouldn't this test be made after submitted as well? Also you should also test whether the other total counts are zero or not.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/StreamingListenerSuite.scala:
```diff
@@ -61,6 +63,32 @@ class StreamingListenerSuite extends TestSuiteBase with Matchers {
+
+    // SPARK-6766: batch info should be submitted
+    val batchInfosSubmitted = collector.batchInfosSubmitted
+    batchInfosSubmitted should have size 4
```
Nit: these tests for submitted and started should come before completed :)

tdas commented:
Just a few minor comments, otherwise LGTM. Can merge as soon as you update these.

zsxwing committed:
Update tests

SparkQA commented:
[Test build #30019 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30019/consoleFull) for   PR 5414 at commit [`2f85060`](https://github.com/apache/spark/commit/2f8506030b572a4702991fd246bdcdf21d7ef509).

tdas commented:
The concerned testsuites have passed, I am merging this to master and 1.3 branch.

tdas commented:
I merged it in master, but could not merge it to 1.3 because of conflicts. Could you open another PR to branch 1.3? Would be good to have these bug fixes in 1.3.x

SparkQA commented:
[Test build #30019 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30019/consoleFull) for   PR 5414 at commit [`2f85060`](https://github.com/apache/spark/commit/2f8506030b572a4702991fd246bdcdf21d7ef509).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30019/
Test PASSed.
